lib/common/zstd_internal.h

```diff
@@ -286,6 +286,27 @@ typedef struct {
     FSE_repeat litlength_repeatMode;
 } ZSTD_entropyCTables_t;
 
+struct ZSTD_CCtx_params_s {
+    ZSTD_compressionParameters cParams;
+    ZSTD_frameParameters fParams;
+
+    int compressionLevel;
+    U32 forceWindow;           /* force back-references to respect limit of
+                                * 1<<wLog, even for dictionary */
+    /* Dictionary */
+    ZSTD_dictMode_e dictMode;   /* select restricting dictionary to "rawContent"
```

Comment:
The same question applies here :
Is it a property of the loading operation ?
In which case, would it be better to make it an argument of the load operation ?

Reply:
As discussed : 
with now 2 possible parameters, loading dictionary is an operation featuring 6 different variants.
So it's no longer practical to use function name to distinguish them

Consider a new prototype `ZSTD_CCtx_loadDictionary_advanced()` with these 2 parameters as arguments.
You can keep existing `ZSTD_CCtx_loadDictionary_byReference()`, which presumes a standard value for `dictMode`.

Prefer using `enum` as argument type, as opposed to integer values, since they are more readable.
The tricky point about `enum` is selecting a correct prefix, so that the naming doesn't pollute namespace.
An example : `{ ZSTD_dl_byRef, ZSTD_dl_byCopy } `.

You can then retrofit the relatively similar `ZSTD_createCDict_advanced()` to ensure it uses the same argument types, in order to keep the API consistent.

Reply:
Re `dictMode`: 
- `ZSTD_CCtx_refPrefix` allows the dict mode to be set via `ZSTD_p_dictMode`, which is relevant in `ZSTD_compress_generic` for the non-multithreaded version. The multithreaded version seems to just treat the mode as `ZSTD_dm_auto`. Was this the intended functionality? Allowing the dict mode option here effectively implies adding a "single-use" dictMode parameter when `ZSTD_CCtx_refPrefix` is used.
- The old functionality in `zstdmt_compress` with setting `ZSTD_p_forceRawDict` in `ZSTDMT_compressChunk` seems like it was a no-op since `ZSTD_compressBegin_advanced` just used `ZSTD_dm_auto`. Am I missing something? If not, is it important to set to raw mode?

Reply:
These are very good points @stellamplau , and it's pretty complex to track.

`ZSTD_CCtx_refPrefix()` main objective is to make it possible to compress a content `C` given another content `P`. This content is typically a very similar copy of `C` such as a previous revision, or simply the previous content in a sequential order controlled by the application. This second use case is representative of `ZSTDMT` methodology.

The primary intention is that such content is "anything", meaning not a formatted dictionary. Hence, it should be loaded using mode `ZSTD_dm_rawContent`. Following the code flow, we can see it ends up in `ZSTD_compress_insertDictionary()`, which accepts `ZSTD_dictMode_e dictMode` as argument, transmitted from `ZSTD_resetCStream_internal()`. Meaning it's not necessarily `ZSTD_dm_rawContent`.

With the disappearance of `dictMode` from `CCtx`, it becomes a problem. Which parameter to provide ?
Well, for `ZSTD_CCtx_refPrefix()`, a quickfix could be to always load in `ZSTD_dm_rawContent` mode. It closes the other mode, but I still have to find a use case which actually needs that.

A complexity is that `ZSTD_compress_usingDict()` also uses a "throw-away" dictionary, but is supposed to load it in `ZSTD_dm_auto` mode. And they currently share the same loading pipeline. Something to look at.

Reply:
Part 2 : It seems it was a broken behavior.
`ZSTD_dm_rawContent` mode was created with `zstdmt` in mind.
`zstdmt` loads a part of previous job as "dictionary", but we want to load it as pure content, with no specific metadata. That's what happen with default `ZSTD_dm_auto`, as long as the content doesn't start with the same magic number as a dictionary built with `zstd`.
The chances of this happening are very small (1 in 4 Billions if values are evenly distributed, but since they are not, and since the magic number is selected to avoid any kind of pattern, it's likely even less). So in practice `ZDSTD_dm_auto` is a very close cousin.
But still, what we want is `ZSTD_dm_rawContent`.

note : we should have a test to dictionary loading modes.
Check that, when dictionary loading is set with `ZSTD_dm_rawContent` mode, when presented with a content starting with the same magic number as a dictionary, it nonetheless load it as raw content.
Method : take a random content, overwrite first bytes with magic number, load it with `ZSTD_dm_auto`, it should fail. Now with `ZSTD_dm_rawContent`, it should work.